### ISSUE TITLE
Add feature to be able to disable pull requests on a repository

### ISSUE THREAD

**Issue by mnemonicflow** [dear-github/dear-github#84]
Like disabling Wiki and Issues in the project's Settings view I would like to be able to disable pull requests on a repository.

This will be very useful for projects which have only a mirror on GitHub and they *DON'T* want to use GitHub's pull request mechanism for accepting contributions.
Or a project open sourced by a company is in the process of accepting GitHub pull requests but not yet, they still have some work to do internally before they are ready. (e.g. Google's TensorFlow project used to [not accept pull requests through GitHub](https://github.com/tensorflow/tensorflow/pull/60#issuecomment-155258983) at the very beginning, but [now they do](https://github.com/tensorflow/tensorflow/pull/728#issuecomment-170106884))

Mirror examples:
- [Linux kernel](https://github.com/torvalds/linux) - 92 people wasted their time sending pull requests because they haven't read how to contribute to the Linux kernel guide
- [Qt project](https://github.com/qtproject) - GitHub is only used as a mirror and not accepting contributions through it, they use their own Gerrit instance
- [kdeconnect-kde](https://github.com/albertvaka/kdeconnect-kde) - GitHub is only used as a mirror and not accepting contributions through it
- [kdeconnect-android](https://github.com/albertvaka/kdeconnect-android) - GitHub is only used as a mirror and not accepting contributions through it

And many others... the maintainer(s) of the project have to tell people every time that submitting pull requests through GitHub is NOT the way to contribute to the project, and to keep telling them to read the contributing guidelines and close their pull request: 
- https://github.com/qtproject/qtbase/pull/6

Having the option to disable pull requests on a project would solve this problem.

**Comment by arekkas** [comment]
Wouldn't this be covered by this?

> Issues and pull requests are often created without any adherence to the CONTRIBUTING.md contribution guidelines, due to the inconspicuous nature of the “guidelines for contributing” link when creating an issue and the fact that it often contains a lot of information that isn’t relevant to opening issues

**Comment by dkfiresky** [comment]
Should there be a separate class for mirror repositories?

**Comment by cvrebert** [comment]
Surprisingly, https://github.com/isaacs/github/issues doesn't seem to have an issue for this yet (or my search-fu is failing me).

**Comment by Stargator** [comment]
:+1:

**Comment by albfan** [comment]
Look for the solution adopted by @GNOME https://mail.gnome.org/archives/desktop-devel-list/2016-November/msg00043.html

cron to the rescue!

**Comment by Flimm** [comment]
I feel `dear-github` is talking about the concerns of open source maintainers. But this one is a huge concern for drive-by contributors like myself. My profile is full of repos that are forks for one-off contributions, and I can't clearly indicate their second-class nature, nor can I highlight the repos that I do want to welcome contributions to. Being able to disable accepting pull requests would help for this purpose.

**Comment by CAFxX** [comment]
> My profile is full of repos that are forks for one-off contributions, and I can't clearly indicate their second-class nature, nor can I highlight the repos that I do want to welcome contributions to. Being able to disable accepting pull requests would help for this purpose.

Agreed. Ideally, in this case, an option that redirects PRs to the upstream fork would be the best possible outcome.

**Comment by yihui** [comment]
I'm tired of asking users not to submit pull requests to a repo I forked, too, e.g., https://github.com/yihui/roxygen2/pulls

**Comment by AGhasemzadeh** [comment]
I need this feature too.
I think it is good.

**Comment by clarkbw** [comment]
Forked repos should likely be handled differently than the original. If you just want to contribute forking seems too heavy weight.

**Comment by ehamiter** [comment]
This would be also extremely helpful for a repo set up for a company's coding challenge. I've had issues with candidates in the past putting up a PR to our repo instead of forking ours and putting the PR on *their* repo. Once a PR is created, our repo is effectively spoiled, since any future candidates simply have to look at the closed PRs and see all the solutions people have written in the past. There's no way (that I've found) to remove them. Preventing them would be great.

**Comment by ShalokShalom** [comment]
👍 +1

**Comment by ShalokShalom** [comment]
@ehamiter The only way that I know is that you simply mirror here and host somewhere else. 
First, on Gitea you can hide the status and second, mirror Github instances cant accept Pull-Requests.

**Comment by ve3** [comment]
I want this too.

**Comment by TomasVotruba** [comment]
This would be awesome for all monorepos that are on the rise in PHP worlds in last 2 yeras

**Comment by codecat** [comment]
I don't understand why there's an option to disable issues, but no option to disable PR's...?

**Comment by mloskot** [comment]
Because PRs are central to GitHub flow, and GitHub encourages to follow the fork & pull request flow. That clearly influences or affects the GitHub UI. That's my take on understanding this issue and GitHub position.

**Comment by dessant** [comment]
I've made a new app that immediately closes and locks new and existing issues or pull requests and also supports posting a comment and labeling. It is perfect for forks and mirrors, and you can configure it to your liking. Let me know how it works for you!

https://github.com/dessant/repo-lockdown

**Comment by codecat** [comment]
Nice! I'll definitely look into this. Not sure if I can add a config file to my mirror repos though.

**Comment by dessant** [comment]
@codecat, the config file can also be added to special repository named `.github`, check the README for details.

**Comment by codecat** [comment]
Thanks. Not sure what you mean with a special "repository" though. Do you mean a branch?

**Comment by dessant** [comment]
No, a regular repository. It's just treated specially by apps using the [probot-config](https://github.com/probot/probot-config) extension. :yum:

**Comment by Schachte** [comment]
I also added a new app similar to this idra. Still adding finishing touches, but you can set Max number of pull requests to be opened at a given time. For this case, setting the max to zero makes it effectively achieve the result you're looking for. Check out WIPCream on my profile.

**Comment by ifprado** [comment]
Can this be a new feature please?

**Comment by ShalokShalom** [comment]
Why does this take so long?

**Comment by atull-altera** [comment]
Yes, need this feature.

**Comment by olegchir** [comment]
Yes, need this feature

**Comment by nmeum** [comment]
Wrote myself a webhook for dealing with this issue https://github.com/nmeum/noprs in case this is useful for someone else as well.

**Comment by ShalokShalom** [comment]
Github does obviously want to keep this as it is currently. 

Just think about it: To switch to another source hosting solution becomes easier, once this is possible.

**Comment by Chris2011** [comment]
It would be cool to block PRs against a specific Branch (e.g. master). Also not opening the PR and auto closing, it just shouldn't be possible to create a PR for that branch

**Comment by mloskot** [comment]
@Chris2011 Good idea. It could be part of branch protection settings.

**Comment by rseichter** [comment]
I just ran into this issue with a new repository, although I clearly documented that GitHub is a read-only mirror. Please add a feature to prevent the creation of pull requests. Thank you.

**Comment by eugenekunkel** [comment]
This feature is badly needed. I've refrained from releasing a large project of mine on Github because I don't want to accept pull requests during the alpha and beta phases of the project, and don't want to waste people's time trying to submit them. 😐

**Comment by dss-fed** [comment]
Another +1 for this feature please. Currently faced with having to delete and recreate a repo or email GitHub support in order to work around this.

**Comment by unclehowell** [comment]
![1f44d](https://user-images.githubusercontent.com/10119070/82590225-59493f80-9b6b-11ea-922d-c81ac7109a4e.png) +1

**Comment by yelizariev** [comment]
A possible workaround could be a Github Action, that runs on every new pull request, posts some message and closes the pull request

**Comment by ljharb** [comment]
That is viable but very hostile compared to preventing the effort from being put in in the first place.

**Comment by dgw** [comment]
Yes, automatically closing unwanted PRs is a kludge, not a workaround.

**Comment by omnipitous** [comment]
I'll join the party... +11!  A whole thread of people who seem to agree we need the feature.
...and yes "auto-closing and nuking PRs" is NOT an acceptable work-a-round for our purposes. We need the PR to never have happened to prevent CI/CD triggers from firing.

**Comment by FriedrichWeinmann** [comment]
> We need the PR to never have happened to prevent CI/CD triggers from firing.

Heck! We need to prevent them just to prevent community frustration/attrition.

Definitely need a way to do this in the repository settings: Who is allowed to PR?

+ Nobody
+ Owner Only
+ Contributors
+ Everybody

Imagine a situation where you want to keep your code public and accept issues, but for legal reasons cannot accept code changes from outside (e.g.: Contributor License Agreement not yet done on a new project)

**Comment by ChrisAcrobat** [comment]
When was the latest official response to this feature? The latest I have seen is [2018](https://stackoverflow.com/a/48940685).

**Comment by SoftCreatR** [comment]
Well, [this GH action](https://github.com/marketplace/actions/repo-lockdown) is a well documented and working solution, until there's something official.

**Comment by ZachSAEON** [comment]
@SoftCreatR, one reason that someone might want to disable pull requests is that a pull request can trigger an action on a public repository. I don't think that closing pull requests immediately after they are created will help with this

**Comment by SoftCreatR** [comment]
> @SoftCreatR, one reason that someone might want to disable pull requests is that a pull request can trigger an action on a public repository. I don't think that closing pull requests immediately after they are created will help with this

Possibly. But in this case, another action might be able to cancel the execution, if a PR get's closed.

**Comment by NilsRo** [comment]
> 
> Definitely need a way to do this in the repository settings: Who is allowed to PR?
> 
> * Nobody
> * Owner Only
> * Contributors
> * Everybody
> 
> Imagine a situation where you want to keep your code public and accept issues, but for legal reasons cannot accept code changes from outside (e.g.: Contributor License Agreement not yet done on a new project)

That’s exactly we need for Github Enterprise to control the developer workflow. Hopefully this will be added sometimes until then we will try to add this with hooks. We will see...

**Comment by soenkeliebau** [comment]
> Well, [this GH action](https://github.com/marketplace/actions/repo-lockdown) is a well documented and working solution, until there's something official.

There is just one issue with that solution, when this action becomes active someone already put effort into creating a pull request, forking the repo, fixing code, writing docs, ... and all that basically just to hear "thanks, but no thanks".
Granted, it should be clear from Readme etc. that pull requests are not welcome, but still, this is not ideal.

**Comment by astrolemonade** [comment]
Status Linux repository:
![image](https://user-images.githubusercontent.com/35623921/110794685-a9ba9680-827e-11eb-9973-968954181635.png)
761 people opened PRs, the Linux team is not even trying to close any of the new ones since it is cumbersome.

**Comment by b1f6c1c4** [comment]
I noticed recently that the Linux repo is no longer allowed to create PRs, with a warning like this:

```
An owner of this repository has limited the ability to open a pull request to users that are collaborators on this repository.
```

Try the following:

https://github.com/torvalds/linux/compare/master...Rust-for-Linux:rust

**Comment by eugenekunkel** [comment]
> I noticed recently that the Linux repo is no longer allowed to create PRs, with a warning like this:
> 
> ```
> An owner of this repository has limited the ability to open a pull request to users that are collaborators on this repository.
> ```
> 
> Try the following:
> 
> https://github.com/torvalds/linux/compare/master...Rust-for-Linux:rust

That is interesting and I can't figure out how to duplicate that behavior. Experimental feature perhaps? After all, Linus Torvalds is the guy who created Git in the first place...

**Comment by SoftCreatR** [comment]
Shouldn't it be this one?

![image](https://user-images.githubusercontent.com/81188/114829217-17e90080-9dcb-11eb-9dd4-69834e2b223e.png)

**Comment by rseichter** [comment]
@SoftCreatR As "interaction" includes creating issues and more, the existing moderation feature cannot be used to disable only pull requests.

**Comment by am11** [comment]
+1, we just need a `Pull Request` checkbox under `https://github.com/:owner:/:repo:/settings > Features`

**Comment by acmnpv** [comment]
This happens from time to time as well for the https://github.com/gromacs/gromacs repo that is just a plain mirror. Please add the ability to disable pull requests for those situations.

**Comment by jantajov** [comment]
> +1, we just need a `Pull Request` checkbox under `https://github.com/:owner:/:repo:/settings > Features`

I wonder how difficult would be to add this feature. This issue is opened for five years and there is still no such checkbox.

**Comment by vincerubinetti** [comment]
One other possible use case I haven't seen mentioned yet is this: [people opening PRs against the upstream repository when they meant to open one against their fork of it](https://github.community/t/changing-the-default-pr-target-when-creating-a-pr-from-a-fork/2842).

I made a website template that people can fork to make their own website. When they go to update their own website through a PR, they accidentally target the template itself rather than the master/main branch of their fork of it. I get several of these accidental PRs a month.

It'd be great to be able to limit the opening of PRs to just certain contributors, the same way as any other action (push, manage issues, etc.) have tiers of permissions in the "Manage access" settings tab.

---

Though it's been said many times many ways, I'll add my further two cents of frustration.

I get the reasoning behind wanting to keep software open to improvement, I truly do. I believe in FOSS. I get wanting to encourage people to improve other peoples' software. But isn't this going overboard. Things are a bit different than they were years ago; more people are contributing. In addition to all the 👍'ed reasons above, do we really need to enforce this so strictly anymore? Can't it just be on by default, like issues, and "strongly recommended", or something like that?

Also... if this were about fostering open source, why is there an option to disable issues, but not pull requests... Now I'm starting to suspect this is just stubbornness and unwillingness to change.

**Comment by andrewfrueh** [comment]
Yup, I need to throw down here as well.

Per the enormous string of previous comments, I would also like to be able to *disable pull requests* on my repos. This is definitely at a point where none of us understands why this is being withheld. GitHub... you need to listen to your community!

I really like [this news item I heard on Changelog about Ben Johnson's statement](https://changelog.com/news/open-source-but-closed-to-contributions-jGyl). To quote:
>I’ve made the decision to keep this project closed to contributions for my own mental health and long term viability of the project.

This is NOT about restricting open-source. This is about the mental health of developers. We only have so much bandwidth.

**Comment by astrojuanlu** [comment]
Beyond lacking the necessary granularity, the "temporary interaction limits" mentioned in https://github.com/dear-github/dear-github/issues/84#issuecomment-820179028 are, well, temporary.

**Comment by Pharap** [comment]
This would be a very welcome feature, but I would add the caveat that it should be possible to instead restrict PRs to being a maintainers only feature.

Often people make a PR that solves their immediate problem without considering the needs of others or the wider implications of their changes. Maintainers understand those needs and implications much better than drive-by contributors, so it makes sense to be able to restrict contributions to only maintainers.

A workflow that consists of users making feature requests and bug reports via issues and then maintainers deciding how best to implement those features or fix those bugs will often run more smoothly than users contributing a half-baked solution and maintainers having to gently explain why they aren't going to accept the PR without seeming like the villain of the story.

**Comment by ADTC** [comment]
@acmnpv suggest gromacs to open a PR with the title: **"DO NOT OPEN A PULL REQUEST HERE."** and leave that pull request open. (Ironic, I know. 😂)

Also adding extra information to `README` or `CONTRIBUTING` would help, but I doubt you can do it in gromacs (since it's a mirror).

@dear-github please allow us to disable all three merge strategies, and in doing so, automatically disable the Pull Request tab and related features as well. The only way to merge branches should be by doing so locally and pushing.

### **Open source ≠ Open contribution**

**Comment by grahamperrin** [comment]
https://github.com/orgs/github-community/discussions/8907

**Comment by daztucker** [comment]
I also would like to disable pull requests on a repo, but for a reason I have not seen mentioned yet: security of selfhosted runners. The according to the [action runner docs](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security) the ability to create pull requests on a repo can be used to compromise the security of a runner attached to that repo.

We have a set of self-hosted runners that run tests on various platforms for which there are no hosted runners, and based on the docs we have them attached to a private repo that mirrors the public one.  I would like to be able to disable pull requests on that repo so I can make it and thus the test statuses public.

(There is an option to disable workflows in pull requests but I'm not sure if that it's sufficient, especially given that the runner docs don't mention it and have only a blanket warning against using selfhosted runners on public repos).

**Comment by Hyaxia** [comment]
For the time being we can block PRs for a limited time using the moderation section

<img width="767" alt="image" src="https://user-images.githubusercontent.com/36747317/180594368-b1cb6402-1cab-40c9-906e-4f1ccc08d348.png">

**Comment by doctorperceptron** [comment]
> For the time being we can block PRs for a limited time using the moderation section
> 
> <img alt="image" width="767" src="https://user-images.githubusercontent.com/36747317/180594368-b1cb6402-1cab-40c9-906e-4f1ccc08d348.png">

As previously pointed out, this isn't granular enough as it will disable _all interactions_, which isn't limited to PRs only.

**Comment by daztucker** [comment]
> For the time being we can block PRs for a limited time using the moderation section

That can only be enabled for a limited time (up to 6 months) so for my use case, having a security control that fails open after some period is problematic.

**Comment by mosesoak** [comment]
Our use case for this: a sandbox repo where we _only_ build one-off demo branches and never merge to main.

**Comment by xCENTx** [comment]
Disabling Pull Requests should be a thing. Im tired of getting harrassed by people who want to spell check my work so they can say they made something I made

**Comment by somebody-new** [comment]
Another example: https://github.com/PySimpleGUI/PySimpleGUI does not take pull requests because it's the best way for them to be able to keep working on the project. At least once a month they have to tell people that the project doesn't take pull requests.

Some people want their work to be publicly available without having to deal with the exhausting job of reviewing other people's code.

**Comment by grahamperrin** [comment]
> exhausting

I'm only _slightly_ tired after reading the title bar of the front page of the project: 

https://user-images.githubusercontent.com/192271/202830440-b7b01ab4-182e-4504-848e-95220c13a99f.png

**Comment by MahaSPR** [comment]
Is there a way to disable pull requests on a particular branch to the main/master branch? If so, please help me how to configure this, thanks in advance.

**Comment by whitespacecode** [comment]
> Is there a way to disable pull requests on a particular branch to the main/master branch? If so, please help me how to configure this, thanks in advance.

Looking for this to. We want to have the development branch to be without pr

**Comment by ghost** [comment]
This is still relevant and clearly is an in demand feature. Really trying to stay optimistic for this one....

**Comment by ugreg** [comment]
This workaround could be useful until there support in the repo settings to disable PRs.

https://github.com/marketplace/actions/close-pull-request

> This Action subscribes to pull_request_target events. When receiving a pull_request_target event, this action closes the pull request triggered by the event immediately.

**Comment by kennyma-eplus** [comment]
> This workaround could be useful until there support in the repo settings to disable PRs.
> 
> https://github.com/marketplace/actions/close-pull-request
> 
> > This Action subscribes to pull_request_target events. When receiving a pull_request_target event, this action closes the pull request triggered by the event immediately.

But if the repo is a code challenge, the "answer" will keep in "closed pull request"

**Comment by karlaxwell** [comment]
> This will be very useful for projects which have only a mirror on GitHub and they DON'T want to use GitHub's pull request mechanism for accepting contributions.

I would be happy about this feature, this case you described is the same for me.

**Comment by ngdangtu-vn** [comment]
Wake me up when the feature arrives.

**Comment by ADTC** [comment]
> code challenge

In this case, it's not enough to disable pull requests. You'd have to disable forking entirely.

Use a private repo, and disable forking on it. Invite only the challenged to view the repo. Publish the repo only after the challenge is complete. Or IDK, maybe you have something else in mind. Is this a challenge that's open to the whole world?

Other services might suit this use case better actually.

PS: Dear GitHub, please add feature to disable pull requests on a repository. You can literally just put a checkbox in Settings > Features and hide the tab if checked...

**Comment by ljharb** [comment]
In that case, you'd have to disable cloning also, which is impossible. However, there's tons of other valid use cases to disable PRs on a repo, so it doesn't really matter that that single use case doesn't really work.

**Comment by quanah** [comment]
The OpenLDAP project mirror is another case where we do not want PR's to be made via GitHub.

**Comment by a-n-d-a-i** [comment]
bump

**Comment by ryubal-sh** [comment]
It's only been 8 years, c'mon we can do this guys

**Comment by fatso83** [comment]
![Screenshot_20240821-085901.png](https://github.com/user-attachments/assets/7670d16a-650c-4f64-874d-cff766a58bd3)

From isaacs/GitHub

Just use the official channel

**Comment by fatso83** [comment]
Upvote the feature here

https://github.com/orgs/community/discussions/8907

**Comment by PaulNewton** [comment]
Currently relevant "Exploring Solutions to Tackle Low-Quality Contributions on GitHub"
https://github.com/orgs/community/discussions/185387

**Comment by younsl** [comment]
Hi, this feature has finally arrived! GitHub has introduced [new repository settings for configuring pull request access](https://github.blog/changelog/2026-02-13-new-repository-settings-for-configuring-pull-request-access/).
